### ISSUE TITLE
Make watch-cache size dynamic and get rid of magic constants

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#90058]
Currently, capacity for watchcache is a parameter at creation time and cannot be changed later:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L81
To make it scalable, we have a number of magic constants:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/cachesize/cachesize.go#L25
and we already had a bunch issues with it (people don't know it should be updated for some new resources, we can't clearly say how to set it, ...)

We should try to get rid of of all those magic constants completely.

`At the high-level,  instead of providing magic constants, we want to change it so that last X minutes of history will be kept (e.g. X=5)`

How can we do that:
1. the first step is to allow modifying size of watchcache in a running apiserver:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L127
- fortunately it's not that hard - it's enough to simply (under lock) create a new array, copy contents and adjust startIndex and endIndex appropriately
2. the second step is decide when the size should be adjusted
- objects don't have information about their last update time, but fortunately we don't need it - we don't have to be super accurate; we can simply assume that the last update time of the object is the time when we learned about it (received it from etcd); we know that the object was modified earlier that that, so if we ensure that we have at least X minutes of history based on those timestamps, we also will have at least X minutes of history based on real timestamps;
- how the above translates to concreate changes? we need to extend watchCacheEvent to also keep the timestamp, and on adding new item to cache, if the item we would have to overwrite is to fresh, trigger resizing of cache (increase by 2x)
3. the remaining part is to decide when we need to reduce the size of cache; we can probably start without that at all, but ideally, we should decrease the size by 2x when the X minutes history would fit into half of it (so 1/4th of the original size) - this should allow us to avoid flapping

@kubernetes/sig-scalability-feature-requests @mm4tt - FYI

**Comment by wojtek-t** [comment]
I'm happy to provide guidance if someone is interested in doing that.

/help wanted

**Comment by lavalamp** [comment]
@deads2k may have thoughts

IMO we need protection against exceeding memory limitations, but the proposal above sounds unbounded?

**Comment by wojtek-t** [comment]
> IMO we need protection against exceeding memory limitations, but the proposal above sounds unbounded?

Sure - we should have some upper bound. But that's one additional "if" and doesn't change anything significantly.

**Comment by lavalamp** [comment]
This got me thinking about the watch cache in general. I wrote up some thoughts here: https://github.com/kubernetes/kubernetes/issues/90179

**Comment by wojtek-t** [comment]
One thing to point out here - my motivation here isn't decreasing the memory footprint (though it will be one of the consequences). My motivation is maintainability and unnecessary complexity and magic constants...

**Comment by lavalamp** [comment]
Oh, you expect this will reduce memory usage? I expected it to increase memory usage, at least in big clusters with high-churn collections.

**Comment by wojtek-t** [comment]
> Oh, you expect this will reduce memory usage? I expected it to increase memory usage, at least in big clusters with high-churn collections.

My expectations is that it will decrease for biggest and smallest cluster. The constants for things like endpoints or pods are huuuge now, e.g.:
- pods: maxInt(50*clusterSize, 1000) [5k nodes => 250.000]
- endpoints:  maxInt(10*clusterSize, 1000) [5k nodes => 50.000]

We may increase in things like nodes or leases:
- nodes: maxInt(5*clusterSize, 1000) [5k nodes => 25.000]

but that should be compensated by descrease in the above ones. I agree that we may increase for things like ReplicaSets or Deployments where there can be a bunch of status updates (and we use default = 100 for them), but that also shouldn't be that huge increase.

I'm still thinking about the constants - I initially proposed 5 minutes, but maybe it's actually unnecessarily long? Maybe 2-3 minutes is enough? Especially if we combine it with adjusting watch bookmarks policy: https://github.com/kubernetes/kubernetes/issues/90160

**Comment by lavalamp** [comment]
Oh, I didn't realize we did it that way, I thought it was a flat value like etcd's watch window. So, yeah, it seems only super-high churn clusters would see an increase.

**Comment by wojtek-t** [comment]
> So, yeah, it seems only super-high churn clusters would see an increase.

Yeah - super high, and not the largest one - probably the medium ones (couple hundreds of nodes - that's my prediction).

**Comment by wojtek-t** [comment]
The main PR has merged - we will be monitoring our scale tests in the next week or so, maybe two, and assuming everything will look fine, we should get rid of the magic constants completely.

But let's wait with it for some time.

**Comment by answer1991** [comment]
Should we remove kube-apiserver flag `--watch-cache-sizes`? After #90091 merged, flag `--watch-cache-sizes=` looks like meaningless.

**Comment by wojtek-t** [comment]
We're not ready. As mentioned above:
- I would like to wait for some time to validate it - I see a little bit more increases than I would expect
- the constants still determine the initial size

Yes - we should remove it, but not now.

**Comment by wojtek-t** [comment]
> Should we remove kube-apiserver flag --watch-cache-sizes? After #90091 merged, flag --watch-cache-sizes= looks like meaningless.

Actually - update on it.
We can't really get rid of this flag, as it allows you to switch on/switch off watch cache for particular resources. We will need to replace it with the one that would allow it.
